src/chainstate/stacks/boot/pox-2.clar

```diff
@@ -0,0 +1,792 @@
+      (map-set stacking-state
+        { stacker: stacker }
+        { amount-ustx: amount-ustx,
+          pox-addr: pox-addr,
+          first-reward-cycle: first-reward-cycle,
+          lock-period: lock-period })
+
+      ;; return the lock-up information, so the node can actually carry out the lock. 
+      (ok { stacker: stacker,
+            lock-amount: amount-ustx,
+            unlock-burn-height: unlock-burn-height })))
+
+;; Reject Stacking for this reward cycle.
+;; tx-sender votes all its uSTX for rejection.
+;; Note that unlike PoX, rejecting PoX does not lock the tx-sender's
+;; tokens.  PoX rejection acts like a coin vote.
+(define-public (reject-pox)
+    (let (
+        (balance (stx-get-balance tx-sender))
+        (vote-reward-cycle (+ u1 (current-pox-reward-cycle)))
+    )
+
+    ;; tx-sender principal must not have rejected in this upcoming reward cycle
+    (asserts! (is-none (get-pox-rejection tx-sender vote-reward-cycle))
+        (err ERR_STACKING_ALREADY_REJECTED))
+
+    ;; tx-sender can't be a stacker
+    (asserts! (is-none (get-stacker-info tx-sender))
+        (err ERR_STACKING_ALREADY_STACKED))
+
+    ;; vote for rejection
+    (map-set stacking-rejection
+        { reward-cycle: vote-reward-cycle }
+        { amount: (+ (next-cycle-rejection-votes) balance) }
+    )
+
+    ;; mark voted
+    (map-set stacking-rejectors
+        { stacker: tx-sender, reward-cycle: vote-reward-cycle }
+        { amount: balance }
+    )
+
+    (ok true))
+)
+
+;; Used for PoX parameters discovery
+(define-read-only (get-pox-info)
+    (ok {
+        min-amount-ustx: (get-stacking-minimum),
+        reward-cycle-id: (current-pox-reward-cycle),
+        prepare-cycle-length: (var-get pox-prepare-cycle-length),
+        first-burnchain-block-height: (var-get first-burnchain-block-height),
+        reward-cycle-length: (var-get pox-reward-cycle-length),
+        rejection-fraction: (var-get pox-rejection-fraction),
+        current-rejection-votes: (next-cycle-rejection-votes),
+        total-liquid-supply-ustx: stx-liquid-supply,
+    })
+)
+
+
+(define-public (stack-extend (extend-count uint)
+                             (pox-addr { version: (buff 1), hashbytes: (buff 20) }))
+   (let ((stacker-info (stx-account tx-sender))
+         (amount-ustx (get locked stacker-info))
+         (unlock-height (get unlock-height stacker-info))
+         (first-extend-cycle (burn-height-to-reward-cycle unlock-height))
+         (last-extend-cycle  (- (+ first-extend-cycle extend-count) u1))
+         (cur-cycle (current-pox-reward-cycle))
+         (lock-period (- last-extend-cycle cur-cycle))
+         (new-unlock-ht (reward-cycle-to-burn-height (+ u1 last-extend-cycle))))
+
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+                (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; tx-sender must be locked
+      (asserts! (> amount-ustx u0)
```

Comment:
Shouldn't we also be checking that this meets the minimum amount?  I don't see a call to `can-stack-stx`, for example.

Reply:
More generally -- do we want to permit someone to `stack-extend` their locked STX indefinitely, even if the quantity stacked becomes smaller than the minimum amount?  I don't think it matters much either way from a technical standpoint, but it does affect how the user finds out that they won't clinch a reward address -- they can either get an error here, or they can wait until a reward cycle completes where they have no reward slots.

Reply:
I'm not sure if it ultimately makes much difference in terms of alerting the user -- the `can-stack-stx` check is performed at the time of the extension, so if the liquid supply of STX increases during their stacking (e.g., they stack for 12 cycles, but the supply increases during cycle 6), they wouldn't be alerted either way.

The biggest difference is in how the user needs to deal with the consequences of this. In any event, to meet the new threshold, they will need to invoke `stack-increase` to increase their lock and stacking amount. However, if the `can-stack-stx` check is in place, then it imposes an ordering constraint: they would need to invoke `stack-increase` before extending.

I think it probably *does* make sense to add the check here, as it seems the least surprising behavior.

Reply:
Added `can-stack-stx` and `minimal-can-stack-stx` checks to `stack-extend` and `delegate-stack-extend` respectively.